compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -3497,6 +3504,39 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
         }
     }
 
+    /// Suggests borrowing if there exists a viable impl for the `trait_pred`'s
+    /// dereferenced self type.
+    /// Although it's not specific, this method targets `.into()` conversions.
+    fn suggest_borrowing_for_method_call(
+        &self,
+        obligation: &PredicateObligation<'tcx>,
+        err: &mut Diagnostic,
+        trait_pred: ty::PolyTraitPredicate<'tcx>,
+    ) {
+        let ty::Ref(_, ty, mutability) = trait_pred.self_ty().skip_binder().kind() else { return };
+
+        let ObligationCauseCode::FunctionArgumentObligation { call_hir_id, .. } = obligation.cause.code() else { return };
+        let Some(hir::Node::Expr(hir::Expr {
+            kind: hir::ExprKind::MethodCall(_, receiver, _, _),
+            ..
+        })) = self.tcx.hir().find(*call_hir_id) else { return };
+
+        let new_obligation = self.mk_trait_obligation_with_new_self_ty(
+            obligation.param_env,
+            trait_pred.map_bound(|trait_pred| (trait_pred, *ty)),
+        );
+        if !self.infcx.predicate_must_hold_modulo_regions(&new_obligation) {
+            return;
+        }
+
```

Comment:
actually... this feels wrong to me

##### we check whether the obligation is some `&(mut) <ty>: Trait<..>`

here that is `&Foo: From<u64>`.

##### it then checks whether the obligation cause code is `FunctionArgumentObligation`.

This only gets added in `fn blame_specific_arg_if_possible`, if for some method `foo` there is only one argument, which references a generic parameter of either the method or the trait. This gets the relevant method using an `ExprItemObligation` or `ExprBindingObligation` cause code. This cause code is for the root obligation which is `u64: Into<&Foo>`, not `&Foo: From<u64>` here.

So for the method call `x.into()` we point to the `x` because it's the only argument mentioning `Self`. Note that we point at the argument `x` of the method `into()`, not at the argument for `foo`.

So the `call_hir_id` is the `HirId` of `x.into()` while the `arg_hir_id` is the id of `x` (which is why we've actually added the `FunctionArgumentObligation`).

I am confused here, we emit `FunctionArgumentObligation` for some random `arg_hir_id` without ever checking that the obligation is in any way related to the param of that argument?

##### we then check whether `<ty>: Trait<..>` holds

this checks `Foo: From<u64>` which ends up working out in your test case, but I don't see how that's generally applicable. The fact that `u64: Into<&Foo>` has a nested obligation where the self type is equal to the actual return type of the `42u64.into()` method call pretty much happens by circumstance.

I have to admit that reading through `adjust_fulfillment_error_for_expr_obligation` i am pretty confused how exactly that is supposed to work :thinking:

Reply:
Sorry, I'm not familiar with the codebase, I've tried to implement it based on the instructions provided in the original issue and it's very likely that I'm missing something 😄. As far as I understand from your comments, would adding some checks to make sure we're dealing with `From` trait and `into()` calls solve the problem? Like:
```rs
fn suggest_borrowing_for_into_calls(
    &self,
    obligation: &PredicateObligation<'tcx>,
    err: &mut Diagnostic,
    trait_pred: ty::PolyTraitPredicate<'tcx>,
) {
    // Check if the obligation is in the form of `&T: From<U>`
    let Some(sym::From) = self.tcx.get_diagnostic_name(trait_pred.def_id()) else { return };
    let ty::Ref(_, ty, mutability) = trait_pred.self_ty().skip_binder().kind() else { return };

    // Check if `T: From<U>` holds, so borrowing is a valid suggestion
    let new_obligation = self.mk_trait_obligation_with_new_self_ty(
        obligation.param_env,
        trait_pred.map_bound(|trait_pred| (trait_pred, *ty)),
    );
    if !self.infcx.predicate_must_hold_modulo_regions(&new_obligation) {
        return;
    }

    // Check if the obligation is caused because of a `x.into()` call
    let ObligationCauseCode::FunctionArgumentObligation { call_hir_id, .. } = obligation.cause.code() else { return };
    let Some(hir::Node::Expr(hir::Expr {
        kind: hir::ExprKind::MethodCall(segment, receiver, _, _),
        ..
    })) = self.tcx.hir().find(*call_hir_id) else { return };
    if segment.ident.as_str() != "into" {
        return;
    }

    // Suggest borrowing
    err.span_suggestion_verbose(
        receiver.span.shrink_to_lo(),
        "consider borrowing here",
        format!("&{}", mutability.prefix_str()),
        Applicability::MachineApplicable,
    );
}

```

Reply:
> Sorry, I'm not familiar with the codebase, I've tried to implement it based on the instructions provided in the original issue and it's very likely that I'm missing something smile. 

no worries, given my suggestion your approach is exactly what I would have expected :thinking: 

> As far as I understand from your comments, would adding some checks to make sure we're dealing with `From` trait and `into()` calls solve the problem?

I don't know. I personally feel like `ObligationCauseCode::FunctionArgumentObligation` as is fundamentally broken and don't think it can be correctly used for what you want to do. `FunctionArgumentObligation` does not care about the return type of the method call in any way. I think using it to correctly apply a reference to the returned value is somewhere between "very inconvenient" and "actively counterproductive".

Not sure how to actually implement this in a nice way :thinking: I think having the infrastructure to add a suggestion similar to your approach is very valuable, but setting that up is not trivial and I don't have a fleshed out plan in mind rn.

Reply:
I see, I'm closing this PR for now. Happy to work on this again if you came up with a plan (and some instructions 😄). Do let me know in the original ticket, thanks for your time!